### PR TITLE
Issue#128

### PR DESCRIPTION
**Pull request by duttarnab**
#128 - Windows setup file needed for oxd service
https://github.com/GluuFederation/oxd/issues/128

We have used procrun to create service of oxd-server.
https://commons.apache.org/proper/commons-daemon/procrun.html

Steps to configure:
============
1. Install JRE 1.8 and set JRE_HOME environment variable.
2. Built oxd-server project. 
3. Copy installer directory from  ${OXD_WORKING_DIR}\oxd-server\target\installer to suitable location on disk.
4. Set OXD_HOME as path to installer directory.
5. Change ${OXD_HOME}/conf/oxd-server.yml as required.
6. To install oxd-server double click on ${OXD_HOME}/bin/install.bat file. Check Task Manager --> Services for status of service. You can run end-points on machine.
7. To uninstall oxd-server double click on ${OXD_HOME}/bin/uninstall.bat file.

### CODE REVIEW

File: oxd-server/pom.xml [hunk 1]
```diff
@@ -189,6 +190,32 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-antrun-plugin</artifactId>
```
Comment (yuriyz) [inline oxd-server/pom.xml]: we don't want any installer generation related things here. If you really want to keep it mavenized, probably good idea to keep it in maven profile. So we can activate it when needed.

File: oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java [hunk 2]
```diff
@@ -16,7 +16,12 @@
 
     public static void main(String[] args) {
         try {
-            new OxdServerApplication().run(args);
+            String mode = args[0];
+            if ("stop".equals(mode)) {
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java]: if (args.length > 0 && "stop".equalsIgnoreCase(args[0])) {
    ServerLauncher.shutdown(true);
    return;
}

File: oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java [hunk 3]
```diff
@@ -16,7 +16,12 @@
...
+                ServerLauncher.shutdown(true);
+            } else {
+                new OxdServerApplication().run(args);
+            }
         } catch (Throwable e) {
             LOG.error("Failed to start oxd-server.", e);
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java]: if we trigger stop and it fails, log message will say that it failed to start which is not accurate.

File: oxd-server/pom.xml (hunk 1 above)
Comment (duttarnab) [reply oxd-server/pom.xml]: Created "installer" profile for generating installer directory.

File: oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java (hunk 2 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java]: Changes done.

File: oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java (hunk 3 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/OxdServerApplication.java]: Added below conditional statement in catch block to identify "start" or "stop" of server.

File: oxd-server/pom.xml [hunk 4]
```diff
@@ -21,176 +22,380 @@
...
+                                        <descriptor>src/main/assembly/src.xml</descriptor>
+                                    </descriptors>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
```
Comment (duttarnab) [inline oxd-server/pom.xml]: Created "installer" profile for generating installer directory.

**Comment by duttarnab** [comment]
- To install oxd-server double click on ${OXD_HOME}/bin/install.bat file. Check Task Manager --> Services for status of installed service. You can run end-points on machine.
- To uninstall oxd-server double click on ${OXD_HOME}/bin/uninstall.bat file.

File: oxd-server/pom.xml [hunk 5]
```diff
@@ -21,176 +22,380 @@
         <version>4.0-SNAPSHOT</version>
     </parent>
 
-    <build>
-        <finalName>oxd-server</finalName>
+    <profiles>
+        <profile>
+            <id>built</id>
```
Comment (yuriyz) [inline oxd-server/pom.xml]: we don't need copy/paste default configuration. Defaults leaves as default and it is already activated. Introduce only `windows-installer` profile with activator lets say `-Dbuild-windows-installer=true`

File: oxd-server/pom.xml [hunk 6]
```diff
@@ -13,6 +13,7 @@
         <h2.version>1.4.194</h2.version>
         <jedis.version>2.9.0</jedis.version>
         <dropwizard.version>1.3.1</dropwizard.version>
+        <installer.dir>${project.basedir}/target/installer</installer.dir>
```
Comment (yuriyz) [inline oxd-server/pom.xml]: this should go inside profile

**Comment by yuriyz** [comment]
I meant maybe we can have installer for oxd. Inside we can have checkbox `Install as Service`. Similar to mysql installer.

image

**Comment by duttarnab** [comment]
Opened New PR#
https://github.com/GluuFederation/oxd/pull/359

**Pull request closed without merging**